luarules/gadgets/unit_evolution.lua

```diff
@@ -349,7 +349,8 @@ if gadgetHandler:IsSyncedCode() then
 						Evolve(unitID, evolutionMetaList[unitID].evolution_target)
 					end
 				end
-				if evolutionMetaList[unitID].evolution_condition == "timer_global" and currentTime >= evolutionMetaList[unitID].evolution_timer then
+				-- evolutionMetaList[unitID] might have been invalidated if Evolve() was called before this, since that will destroy the unit.
+				if evolutionMetaList[unitID] and evolutionMetaList[unitID].evolution_condition == "timer_global" and currentTime >= evolutionMetaList[unitID].evolution_timer then
```

Comment:
The inside of the `if` is the same in both cases, perhaps just do a single check with an OR?

Reply:
sorry, not sure what you mean... the check I'm doing there should protect the whole `if` clause, the one a few lines earlier doesn't have the problem since it's the Evolve at line 349 that will invalidate the table position if it gets called..

Reply:
I mean that instead of `if A then foo; if B then foo` you could have `if A or B then foo`:
![image](https://github.com/user-attachments/assets/d24925a8-a8bb-4647-a12d-ab393e582878)

Reply:
Yes, but the first block doesn't have the problem because always runs with a fresh unitID.

It's after the first Evolve() call where it can fail, and that's the second test only.

I can add the test also on the first if if that makes you feel better but technically it's not needed, at least not from the report we have, where its failing on the second test.

I was initially baffled too, like... why is this failing on the second test if the first one does the same? but it turns out the unitID is destroyed in Evolve() thus invalidating the array item at the UnitDestroyed callback.

Reply:
Done.